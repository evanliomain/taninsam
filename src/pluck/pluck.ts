/**
 * @module object=>object
 */
/**
 * Create a new object from an object, with a subset of specified properties
 * @param properties The subset of properties
 * @return the function to apply on the object to create its subset of properties
 * @example
 * ```
 * pluck<{ a: number; b: string; c: boolean }, 'a' | 'c'>(
 *    { a: 1, b: 'B', c: true }
 * )('a', 'c') // { a: 1, c: true }
 * ```
 * @example Using the chain
 * ```
 * chain({ a: 1, b: 'B', c: true })
 *   .chain(pluck<{ a: number; b: string; c: boolean }, 'a' | 'c'>())
 *   .value() // { a: 1, c: true }
 * ```
 */
export function pluck<T, K extends keyof T>(
  ...properties: K[]
): (obj: Readonly<T>) => Readonly<{ [key: string]: T[K] }> {
  return (obj: T): Readonly<{ [key: string]: T[K] }> =>
    properties
      .map((property: K) => ({ [property]: obj[property] }))
      .reduce((result, o) => ({ ...result, ...o }), {});
}
