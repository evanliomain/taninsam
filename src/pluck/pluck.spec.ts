import { pluck } from './pluck';

describe('pluck function', () => {
  const input = { a: 1, b: 'B', c: true };

  test('is a pure function', () => {
    pluck()(input);
    expect(input).toMatchSnapshot();
  });

  test("{ a: 1, b: 'B', c: true } |> pluck('a', 'c') === { a: 1, c: true }", () => {
    expect(
      pluck<{ a: number; b: string; c: boolean }, 'a' | 'c'>('a', 'c')(input)
    ).toMatchSnapshot();
  });
});
