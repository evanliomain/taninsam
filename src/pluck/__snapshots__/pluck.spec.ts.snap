// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pluck function { a: 1, b: 'B', c: true } |> pluck('a', 'c') === { a: 1, c: true } 1`] = `
Object {
  "a": 1,
  "c": true,
}
`;

exports[`pluck function is a pure function 1`] = `
Object {
  "a": 1,
  "b": "B",
  "c": true,
}
`;
